### PR TITLE
VAL: Log loss for evaluation

### PR DESCRIPTION
**Pull request by talafek96**
Dismiss BF16 attempt
Fix bugs and refactor code
Change default temp directory

### CODE REVIEW

File: extern/guided_diffusion/guided_diffusion/image_datasets.py [hunk 1]
```diff
@@ -37,14 +39,19 @@ def load_data(
...
     if not data_dir:
         raise ValueError("unspecified data directory")
     
     classes = None
 
+    torch.manual_seed(42)
+
     if few_shot_count is not None:
-        all_files = _select_count_samples_from(data_dir, few_shot_count, deterministic)
+            all_files = _select_count_samples_from(data_dir, few_shot_count + validation_size, deterministic)
```
Comment (amitir22) [inline extern/guided_diffusion/guided_diffusion/image_datasets.py:54]: remove weird redundant tab.

File: extern/guided_diffusion/guided_diffusion/train_util.py [hunk 2]
```diff
@@ -43,6 +44,7 @@ def __init__(
         self.model = model
         self.diffusion = diffusion
         self.data = data
+        self.val_data: DataLoader = val_data
```
Comment (amitir22) [inline extern/guided_diffusion/guided_diffusion/train_util.py:47]: should we leave this type hinting? maybe consider moving the hints to the arguments instead.

File: extern/guided_diffusion/guided_diffusion/train_util.py [hunk 3]
```diff
@@ -151,6 +153,20 @@ def _load_optimizer_state(self):
...
+        all_losses = []
+        logger.log(f"Evaluating model at step {self.step + self.resume_step}")
+        with th.no_grad():
+            for batch, cond in self.val_data:
+                for loss in self.calculate_losses(batch, cond, should_log=False):
+                    all_losses.append(loss.cpu())
+                # TODO: Evaluate the batch of images and log the resulting figure
+        mean_loss = th.tensor(all_losses).mean().item()
+        logger.log(f"\tMean loss on validation data: {mean_loss:.6f}")
+
```
Comment (amitir22) [inline extern/guided_diffusion/guided_diffusion/train_util.py:169]: fuck off those todos

File: core/base_algo.py [hunk 4]
```diff
@@ -371,16 +371,13 @@ def get_remaining_categories(self) -> Set:
         return ALL_CATEGORIES - categories_in_file
 
     def _update_results_csv(self, values_dict) -> None:
-        print('Entered update results csv')  # TODO: Remove this
```
Comment (amitir22) [inline core/base_algo.py:374]: maybe we should replace prints with logging?
if you want to maybe do it later then perhaps adding a TODO will suffice.

File: extern/guided_diffusion/guided_diffusion/image_datasets.py (hunk 1 above)
Comment (talafek96) [reply extern/guided_diffusion/guided_diffusion/image_datasets.py:54]: Done

File: scripts/train.py [hunk 5]
```diff
@@ -3,47 +3,104 @@
...
-    args_to_dict,
     add_dict_to_argparser,
 )
 from guided_diffusion.train_util import TrainLoop
 from utils.models import ModelLoader
 
 
-def main():
-    args = create_argparser().parse_args()
+DEFAULT_LOG_BASE = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'tmp'))
```
Comment (amitir22) [inline scripts/train.py:23]: move to config/configuration.py?

File: extern/guided_diffusion/guided_diffusion/train_util.py (hunk 3 above)
Comment (talafek96) [reply extern/guided_diffusion/guided_diffusion/train_util.py:169]: Next PR

File: core/base_algo.py (hunk 4 above)
Comment (talafek96) [reply core/base_algo.py:374]: No, I have no idea why this even stayed in our code

File: scripts/train.py (hunk 5 above)
Comment (talafek96) [reply scripts/train.py:23]: No, this is specific to this script file

File: utils/models.py [hunk 6]
```diff
@@ -100,9 +100,10 @@ def get_model(self, model_name: str, extra_args: dict=None, to_compile: bool=Tru
         model.eval()
 
         if to_compile:
-            match = re.match(r'([0-9]+\.[0-9]+)', torch.__version__)
-            if match and float(match.group(1)) >= 2.1:  # PyTorch will not support our model before v2.1
-                model = torch.compile(model)
+            # match = re.match(r'([0-9]+\.[0-9]+)', torch.__version__)
+            # if match and float(match.group(1)) >= 2.1:  # PyTorch will not support our model before v2.1
+            #     model = torch.compile(model)
+            pass  # Still unsupported :')
```
Comment (amitir22) [inline utils/models.py:106]: either add another if/flag to the run command, or remove entirely.

Comment (talafek96) [reply utils/models.py:106]: I prefer keeping it so that when in the future it will be supported we will be easily able to put it back in

File: extern/guided_diffusion/guided_diffusion/train_util.py (hunk 2 above)
Comment (talafek96) [reply extern/guided_diffusion/guided_diffusion/train_util.py:47]: I prefer being inconsistently better than consistently bad

**Review by amitir22 — APPROVED** [review]
